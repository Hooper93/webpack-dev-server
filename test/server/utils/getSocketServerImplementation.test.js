'use strict';

/* eslint-disable constructor-super, no-empty-function, no-useless-constructor, no-unused-vars, class-methods-use-this */

const getSocketServerImplementation = require('../../../lib/utils/getSocketServerImplementation');
const BaseServer = require('../../../lib/servers/BaseServer');
const SockJSServer = require('../../../lib/servers/SockJSServer');

describe('getSocketServerImplementation util', () => {
  it("should work with string serverMode ('sockjs')", () => {
    let result;

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: 'sockjs',
      });
    }).not.toThrow();

    expect(result).toEqual(SockJSServer);
  });

  it('should work with serverMode (SockJSServer class)', () => {
    let result;

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: SockJSServer,
      });
    }).not.toThrow();

    expect(result).toEqual(SockJSServer);
  });

  it('should work with serverMode (SockJSServer full path)', () => {
    let result;

    expect(() => {
      result = getSocketServerImplementation({
        serverMode: require.resolve('../../../lib/servers/SockJSServer'),
      });
    }).not.toThrow();

    expect(result).toEqual(SockJSServer);
  });

  it('should throw with serverMode (bad path)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: '/bad/path/to/implementation',
      });
    }).toThrow(/serverMode must be a string denoting a default implementation/);
  });

  it('should throw with serverMode (not extending BaseServer)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: class ServerImplementation {},
      });
    }).toThrow(
      'serverMode must extend the class BaseServer, found via require(webpack-dev-server/lib/servers/BaseServer)'
    );
  });

  it('should throw with serverMode (incorrect constructor)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: class ServerImplementation extends BaseServer {
          constructor() {}
        },
      });
    }).toThrow(
      'serverMode must have a constructor that takes a single server argument and calls super(server)'
    );
  });

  it('should throw with serverMode (no send method)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }
        },
      });
    }).toThrow(
      'serverMode must have a send(connection, message) method that sends the message string to the provided client connection object'
    );
  });

  it('should throw with serverMode (incorrect send parameters)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }

          send(connection) {}
        },
      });
    }).toThrow(
      'serverMode must have a send(connection, message) method that sends the message string to the provided client connection object'
    );
  });

  it('should throw with serverMode (no close method)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }

          send(connection, message) {}
        },
      });
    }).toThrow(
      'serverMode must have a close(connection) method that closes the provided client connection object'
    );
  });

  it('should throw with serverMode (no onConnection method)', () => {
    expect(() => {
      getSocketServerImplementation({
        serverMode: class ServerImplementation extends BaseServer {
          constructor(server) {
            super(server);
          }

          send(connection, message) {}

          close(connection) {}
        },
      });
    }).toThrow(
      'serverMode must have a onConnection(f) method that calls f(connection) whenever a new client connection is made'
    );
  });
});
