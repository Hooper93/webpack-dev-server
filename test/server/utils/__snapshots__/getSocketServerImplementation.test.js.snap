// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSocketServerImplementation util should throw with serverMode (bad path) 1`] = `[Error: serverMode must be a string denoting a default implementation (e.g. 'sockjs'), a full path to a JS file which exports a class extending BaseServer (webpack-dev-server/lib/servers/BaseServer) via require.resolve(...), or the class itself which extends BaseServer]`;

exports[`getSocketServerImplementation util should throw with serverMode (no close, onConnection methods) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options should have required property 'close'
options should have required property 'onConnection'
]
`;

exports[`getSocketServerImplementation util should throw with serverMode (no constructor, send, close, onConnection methods) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options should have required property 'constructor'
options should have required property 'send'
options should have required property 'close'
options should have required property 'onConnection'
]
`;

exports[`getSocketServerImplementation util should throw with serverMode (no onConnection method) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options should have required property 'onConnection'
]
`;

exports[`getSocketServerImplementation util should throw with serverMode (no send, close, onConnection methods) 1`] = `
[ValidationError: webpack Dev Server Invalid Options

options should have required property 'send'
options should have required property 'close'
options should have required property 'onConnection'
]
`;
